src/include/platform/DiagnosticDataProvider.h

```diff
@@ -184,19 +184,12 @@ class DiagnosticDataProvider
 };
 
 /**
- * Returns a reference to a DiagnosticDataProvider object.
+ * Returns the public interface of the DiagnosticDataProvider singleton object.
  *
- * Applications should use this to access the features of the DiagnosticDataProvider.
+ * chip applications should use this to access features of the DiagnosticDataProvider object
+ * that are common to all platforms.
  */
-DiagnosticDataProvider & GetDiagnosticDataProvider();
-
-/**
- * Sets a reference to a DiagnosticDataProvider object.
- *
- * This must be called before any calls to GetDiagnosticDataProvider. If a nullptr is passed in,
```

Comment:
Does this mean applications cannot set a custom DiagnosticDataProvider?

Reply:
The app does not need to setup DiagnosticDataProvider, it should get the DiagnosticDataProvider object implemented by its platform by calling DiagnosticDataProvider::GetDiagnosticDataProvider

"There is a way for applications to set the diagnostic data provider to something of their choice (as now). Probably by handing Server the right pointer."

Who are going to handle over the right pointer to the server? only each platform implementation has this knowledge.

Reply:
Are we sure the default platform implementations are the right ones for all applications running on those platforms?

Reply:
Currently, the implementation of DiagnosticDataProvider class is done by different platform, since some of its APIs are highly platform dependent 

    virtual CHIP_ERROR GetCurrentHeapFree(uint64_t & currentHeapFree);
    virtual CHIP_ERROR GetCurrentHeapUsed(uint64_t & currentHeapUsed);

This PR does not change this design, SetDiagnosticDataProvider is only used by platform layer right now, not application, since application does not have the knowledge of how DiagnosticDataProvider object is substantiated.

Reply:
> Currently, the implementation of DiagnosticDataProvider class is done by different platform, since some of its APIs are highly platform dependent
> 
> ```
> virtual CHIP_ERROR GetCurrentHeapFree(uint64_t & currentHeapFree);
> virtual CHIP_ERROR GetCurrentHeapUsed(uint64_t & currentHeapUsed);
> ```
> 
> This PR does not change this design, the goal of this PR is not to setup the class over platfromMgr. SetDiagnosticDataProvider is only used by platform layer right now, not application, since application does not have the knowledge of how DiagnosticDataProvider object is substantiated.